In [1]:
#Importaciones
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split #Division de los datos
from sklearn.metrics import mean_squared_error #Metrica REgresion Lineal
#Arbol
from sklearn.tree import DecisionTreeRegressor
#Regresion
from sklearn.linear_model import LinearRegression
#kNn
from sklearn import neighbors

In [4]:
df=pd.read_csv("data/df_alumno.csv") #Carga de los datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x1      800 non-null    float64
 1   x2      800 non-null    float64
 2   x3      800 non-null    float64
 3   x4      800 non-null    float64
 4   y       800 non-null    float64
dtypes: float64(5)
memory usage: 31.4 KB


In [5]:
#Division de los datos en x e y previo al armado del train y test
X=df.drop(columns=['y'])

y=df[["y"]]


In [6]:

#Train Test Split
x_train,x_test,y_train, y_test=train_test_split(X,y,  test_size= 0.2, random_state=123) 

In [7]:
#Validation Train Test Split

xval_train,xval_test,yval_train, yval_test=train_test_split(x_train,y_train,  test_size= 0.2, random_state=123)

# Modelos

## Árbol 

In [8]:
#Pruebo diversas profundidades

for depth in range(1,12):
    Mod_Arbol= DecisionTreeRegressor(max_depth=depth, random_state=123 )
    Mod_Arbol.fit(xval_train, yval_train)
    yval_pred= Mod_Arbol.predict(xval_test)
    RMSE= mean_squared_error(yval_test, yval_pred,squared=False)
    print("Arbol de pronfundidad:", depth, "que tiene RMSE:", RMSE,"\n ")
    

Arbol de pronfundidad: 1 que tiene RMSE: 66.9026078871195 
 
Arbol de pronfundidad: 2 que tiene RMSE: 50.57304842041976 
 
Arbol de pronfundidad: 3 que tiene RMSE: 42.088917004388975 
 
Arbol de pronfundidad: 4 que tiene RMSE: 34.743934056322935 
 
Arbol de pronfundidad: 5 que tiene RMSE: 33.98070337158652 
 
Arbol de pronfundidad: 6 que tiene RMSE: 30.86005268719376 
 
Arbol de pronfundidad: 7 que tiene RMSE: 29.410045276470083 
 
Arbol de pronfundidad: 8 que tiene RMSE: 28.519290801672366 
 
Arbol de pronfundidad: 9 que tiene RMSE: 27.22082360150172 
 
Arbol de pronfundidad: 10 que tiene RMSE: 28.361632496846457 
 
Arbol de pronfundidad: 11 que tiene RMSE: 28.367085041353654 
 


## Regresión

In [9]:
Reg_li=LinearRegression()
Reg_li.fit(xval_train, yval_train)
yval_predrl=Reg_li.predict(xval_test)
RMSErl=mean_squared_error(yval_test, yval_predrl,squared=False)
print("El RMSE de Regresion es de:", RMSErl)

El RMSE de Regresion es de: 8.246667760815172e-14


# kNn

In [57]:
#Modelo con peso uniforme

for n in range (1,11):
    Ku_mod=neighbors.KNeighborsRegressor(n_neighbors=n, weights='uniform')
    Ku_mod.fit(xval_train,yval_train)
    yval_predK=Ku_mod.predict(xval_test)
    RMSEk=mean_squared_error(yval_test, yval_predK,squared=False)
    print("Cantidad de vecinos:", n, "y tiene RMSE:", RMSEk,"\n ")

Cantidad de vecinos: 1 y tiene RMSE: 30.970068689261424 
 
Cantidad de vecinos: 2 y tiene RMSE: 28.137926119942833 
 
Cantidad de vecinos: 3 y tiene RMSE: 26.621148477870488 
 
Cantidad de vecinos: 4 y tiene RMSE: 26.632321991619737 
 
Cantidad de vecinos: 5 y tiene RMSE: 25.703577683140903 
 
Cantidad de vecinos: 6 y tiene RMSE: 25.557007693637818 
 
Cantidad de vecinos: 7 y tiene RMSE: 25.796173697959713 
 
Cantidad de vecinos: 8 y tiene RMSE: 25.60000142035879 
 
Cantidad de vecinos: 9 y tiene RMSE: 25.83578151914924 
 
Cantidad de vecinos: 10 y tiene RMSE: 26.497936875679322 
 


In [58]:
#Modelo con peso considerado

for n in range (1,10):
    Kd_mod=neighbors.KNeighborsRegressor(n_neighbors=n,weights='distance' )
    Kd_mod.fit(xval_train,yval_train)
    yval_predD=Kd_mod.predict(xval_test)
    RMSEkd=mean_squared_error(yval_test, yval_predD,squared=False)
    print("Cantidad de vecinos:", n, "y tiene RMSE:", RMSEkd,"\n ")

Cantidad de vecinos: 1 y tiene RMSE: 30.970068689261424 
 
Cantidad de vecinos: 2 y tiene RMSE: 27.605340083793465 
 
Cantidad de vecinos: 3 y tiene RMSE: 25.930096904177702 
 
Cantidad de vecinos: 4 y tiene RMSE: 25.81825784904306 
 
Cantidad de vecinos: 5 y tiene RMSE: 24.996966068897084 
 
Cantidad de vecinos: 6 y tiene RMSE: 24.90545697038405 
 
Cantidad de vecinos: 7 y tiene RMSE: 25.080367239576837 
 
Cantidad de vecinos: 8 y tiene RMSE: 24.966875566985106 
 
Cantidad de vecinos: 9 y tiene RMSE: 25.136049271326538 
 


## Conclusión

El modelo con menor RMSE es el de Regresion Lineal; sin embargo es tan bajo que cae en el overfiting y por eso no lo usaré.
1. El modelo que elijo es el de kNn:
    * Cantidad de vecinos: 6
    * weights='distance'
    * RMSE: 24.90545697038405

# Evaluacion del modelo

In [59]:
#Evaluo en train y testeo en test
Kd_mod=neighbors.KNeighborsRegressor(n_neighbors=6,weights='distance' )
Kd_mod.fit(x_train,y_train)
y_predD=Kd_mod.predict(x_test)
RMSEkd=mean_squared_error(y_test, y_predD,squared=False)
print("El modelo seleccionado tiene RMSE:", RMSEkd,"\n ")

El modelo seleccionado tiene RMSE: 22.503684071807037 
 


- Podemos ver que el RMSE bajó una vez entrenado con los datos del train y evaluados en el test

# Modelo de produccion

Entreno el modelo elegido con TODOS los datos originales

In [34]:
#Evaluo en train y testeo en test
mod_prod=neighbors.KNeighborsRegressor(n_neighbors=6,weights='distance' )
mod_prod.fit(X,y)


KNeighborsRegressor(n_neighbors=6, weights='distance')

### Nuevas predicciones

In [36]:
 #Cargo los datos
df_nuevo= pd.read_csv("data/df_X_reales.csv")
#df_nuevo.info()

In [50]:
#Predigo con el modelo de produccion sobre los valores nuevos y los guardo en la variable
ypred=mod_prod.predict(df_nuevo)
#ypred

In [48]:
#Genero un df con los datos predichos para poder guardarlo en csv
df_ypred = pd.DataFrame(ypred)
#df

In [51]:
#Guardo los datos predichos en el csv
df_ypred.to_csv('Lahoz_Nahuel.csv')